sympy/core/compatibility.py

```diff
@@ -396,16 +396,18 @@ def as_int(n, strict=True):
     ...
     ValueError: ... is not an integer
     """
-    from sympy.core.numbers import Integer
-    try:
-        if strict and not isinstance(n, SYMPY_INTS + (Integer,)):
-            raise TypeError
-        result = int(n)
-        if result != n:
-            raise TypeError
-        return result
-    except TypeError:
-        raise ValueError('%s is not an integer' % (n,))
+    if strict:
+        try:
+            return operator.index(n)
+        except TypeError:
+            raise ValueError('%s is not an integer' % (n,))
+    else:
+        try:
+            if n != int(n):
+                raise TypeError
+            return int(n)
+        except TypeError:
+            raise ValueError('%s is not an integer' % (n,))
```

Comment:
Here we raise TypeError so that we can catch it and then raise ValueError. It would be better just to raise ValueError in the first place and then the outer try is not needed.